python/dcr_template.py

```diff
@@ -230,25 +222,25 @@ def _apply_dcr_kernel(dcr_kernel, model_vals):
         return(np.reshape(template_vals, (size, size)))
 
     @staticmethod
-    def _extract_model_vals(j, i, radius=None, model=None, weights=None):
+    def _extract_model_vals(j, i, model_arr, inverse_weights, radius=None):
         """!Return all pixles within a radius of a given point as a 1D vector for each dcr plane model.
 
         @param j  Vertical pixel index
         @param i  Horizontal pixel index
+        @param model_arr  dcrModel read in with DcrModel._load_model or DcrCorrection, a list of 2D arrays
+        @param inverse_weights   A 2D array containing the inverse of the nonzero elements of the weights
+                                 returned by DcrCorrection, with zeros everywhere else.
         @param radius  radius, in pixels, of the box surrounding (j, i) to be extracted.
-        @param model  dcrModel read in with DcrModel._load_model or DcrCorrection
         @return Returns the weighted model values within the range of pixels, formatted for _apply_dcr_kernel
         """
-        model_arr = []
-        if weights is None:
-            weights = np.ones_like(model)
-        for f in range(model.shape[0]):
-            model_use = model[f, j - radius: j + radius + 1, i - radius: i + radius + 1].copy()
-            weights_use = weights[f, j - radius: j + radius + 1, i - radius: i + radius + 1]
-            model_use[weights_use > 0] /= weights_use[weights_use > 0]
-            model_use[weights_use <= 0] = 0.0
-            model_arr.append(np.ravel(model_use))
-        return np.hstack(model_arr)
+        n_step = len(model_arr)
+        n_pix = (2*radius + 1)**2
+        model_return = np.zeros(n_step*n_pix, dtype=np.float64)
+        inv_weights_use = inverse_weights[j - radius: j + radius + 1, i - radius: i + radius + 1]
```

Comment:
It needs to be a square region centered on the point (j, i), so I'll update the documentation.